src/polkit/polkitunixprocess.c

```diff
@@ -932,6 +927,7 @@ polkit_unix_process_set_pidfd (PolkitUnixProcess *process,
       GError *error;
       error = NULL;
 
+      process->pidfd_is_safe = TRUE;
```

Comment:
This doesn't look right either, the purpose of this is very specifically to know if it came from the dbus peer or not, and it's lost with this change. Please drop it.

Reply:
The check in the `->constructed()` callback never worked in the first place because the setter for the "pid" property also sets the pidfd field, which means if the process was constructed using a "pid" property then by the time `->constructed()` is invoked the pidfd wouls already be set and the pidfd is then incorrectly marked as safe.

Please see https://docs.gtk.org/gobject/vfunc.Object.constructed.html as well as the commit messages for the changes.

Reply:
What tests have you performed to verify those claims, in details?

Reply:
See the code for `polkit_unix_process_set_pid` which is called whenever the "pid" GObject property is set: https://github.com/polkit-org/polkit/blob/894e11120d096b469677830aa94d40e309c22fe6/src/polkit/polkitunixprocess.c#L818-L866

According to the GObject docs linked above, it should be invoked before `->construct()` if the object was constructed with a "pid" property.

Reply:
There's another problem however: all `G_PARAM_CONSTRUCT` properties are set with their default values during construction, and there's no guarantees about the orders in which properties are set. This means if the object was constructed using a pidfd, both the "pid" and "pidfd" properties are set in unspecified order to 0 and the actual pidfd respectively. The same is true for constructing the object from a PID, where the "pidfd" property is also set to -1 which may cause the pidfd set by the "pid" setter to be overwritten. I will push changes to address this soon.

Reply:
Sorry, but the question was intentionally about which tests were performed to verify those claims. I assume this means no tests were done?

I asked about tests because a very simple check shows the claim to be incorrect. The flag is set correctly when the dbus daemon sends a pidfd, and it's not set when it doesn't and it's opened locally instead, as expected and documented.

So once again, please drop this change, and part of the other too, and just do add a check for the ppidfd existence, which will solve any issue with potentially duplicated FDs. Thanks.

Reply:
https://github.com/polkit-org/polkit/blob/894e11120d096b469677830aa94d40e309c22fe6/src/polkit/polkitunixprocess.c#L840

Seems like `polkit_unix_process_set_pid` is using `process->pid` (instead of the `pid` parameter) for `pidfd_open` even if it hasn't been initialized yet. This would obviously result in no pidfd being opened (since 0 is not a valid PID):
```
pidfd_open(0, 0)                        = -1 EINVAL (Invalid argument)
```

Would it be acceptable to just remove pidfd handling from `set_pid` entirely since this code path apparently never worked in the first place?

Reply:
That is needed in some cases, depending on how it's called. Please just fix that issue by changing from process->pid to pid.